smolyay/samples.py

```diff
@@ -0,0 +1,496 @@
+import collections.abc
+import abc
+import numpy
+
+
+class UnidimensionalPointSet(collections.abc.Sequence):
+    """Set of unidimensional points
+
+    A set of unique unidimensional points within a domain.
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+    """
+
+    def __init__(self, domain):
+        self._points = None
+        self._domain = None
+        self._valid_cache = False
+
+        self.domain = domain
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `points`."""
+        return self._domain
+
+    @domain.setter
+    def domain(self, value):
+        domain = numpy.sort(numpy.array(value, dtype=float))
+        if domain.shape != (2,):
+            raise TypeError("Domain must be array with two variables")
+        if domain[0] == domain[1]:
+            raise ValueError("Lower bound must be less than upper bound")
+        if not numpy.array_equal(self._domain, domain):
+            self._domain = domain
+            self._valid_cache = False
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points in the set."""
+        if not self._valid_cache:
+            self._create()
+            self._valid_cache = True
+        return self._points
+
+    def __len__(self):
+        return len(self.points)
+
+    def __getitem__(self, key):
+        return self.points[key]
+
+    @abc.abstractmethod
+    def _create(self):
+        """Create the points in the set."""
+        pass
+
+    def _scale_to_domain(self, points, old_domain):
+        points = self.domain[0] + (self.domain[1] - self.domain[0]) * (
+            (points - old_domain[0]) / (old_domain[1] - old_domain[0])
+        )
+        numpy.clip(points, self.domain[0], self.domain[1], out=points)
+        return points
+
+
+class NestedUnidimensionalPointSet(UnidimensionalPointSet):
+    """Set of unidimensional points assigned to levels.
+
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    num_level: int
+       The number of levels.
+    """
+
+    def __init__(self, domain, num_level):
+        super().__init__(domain)
+        self._num_level = None
+        self._num_per_level = None
+        self._start_level = None
+        self._end_level = None
+
+        self.num_level = num_level
+
+    @property
+    def num_level(self):
```

Comment:
I would make this plural everywhere, since it reads a little strangely as "number of level"
```suggestion
    def num_levels(self):
```